packages/roc-package-web-app-react/app/server/render.js

```diff
@@ -99,13 +96,11 @@ export function reactRender(url, createRoutes, store, renderPage, staticRender =
                     };
                 }
 
-                getPrefetchedData(components, locals)
+                trigger('fetch', components, locals)
                     .then(() => {
-                        let component = <RoutingContext {...renderProps} />;
+                        let component = <RouterContext {...renderProps} />;
```

Comment:
Yeah sorry, I was writing on my phone, so I might have been quite confusing. Ok, so in Hyperion, we have `fetchData` and `fetchDataDeferred`, right? Now actually __both__ of them run on the server as well as the client. When we get a `ROUTER_DID_CHANGE` action, we run the `fetchData`-method. When this action resolves, then we dispatch the original `ROUTER_DID_CHANGE` action and the page renders for the new location. Then `fetchDataDeferred` actions are run, and they in turn can update the state by dispatching actions, causing the page to re-render as needed.

_On the server_:
* We listen to the resolution of the data fetches (a `Promise` is made that resolved once both `fetchData` and `fetchDataDeferred` is complete)
* When the data fetching is complete, we render the page and all data is available

_On the client_:
* No special handling. `@connect` ensures that the page is up to date.

### Pros and cons of this approach
**Pros**
* The article page uses `fetchDataDeferred` to allow us to immediately render the Article page when clicking on a teaser from a section page. This gives the user immediate feedback that something is happening. It also gives us a chance to render a Facebook-style Article placeholder page while the article data is fetched
* We fetch all data needed for rendering on the server
* Allowing the client side to fetch data before and/after router transition is pretty sweet. Perhaps a better name for the methods would be `fetchDataBeforeRouterTransition` and `fetchDataAfterRouterTransition`

**Cons**
* If `fetchDataDeferred` is used, then the route handler must handle the case where the data is not yet available
* If we _actually do want to skip loading_ some data on the server, then we have no good way of triggering client side only data fetches.
* We have no concept of "optional" data that is allowed to fail on the server and retries on the client